Source/Dafny/Compilers/Compiler-Csharp.cs

```diff
@@ -92,6 +92,62 @@ public CsharpCompiler(ErrorReporter reporter)
     }
 
     protected override void EmitBuiltInDecls(BuiltIns builtIns, ConcreteSyntaxTree wr) {
+      EmitInitNewArrays(builtIns, wr);
+
+      EmitFuncExtensions(builtIns, wr);
+    }
+
+    // Generates casts for functions of those arities present in the program, like:
+    //   public static class FuncExtensions {
+    //     public static Func<U, UResult> DowncastClone<T, TResult, U, UResult>(this Func<T, TResult> F,
+    //         Func<U, T> ArgConv, Func<TResult, UResult> ResConv) {
+    //       return arg => ResConv(F(ArgConv(arg)));
+    //     }
+    //     ...
+    //   }
+    // They aren't in any namespace to make them universally available.
+    private static void EmitFuncExtensions(BuiltIns builtIns, ConcreteSyntaxTree wr) {
+      var funcExtensions = wr.NewNamedBlock("public static class FuncExtensions");
+      foreach (var kv in builtIns.ArrowTypeDecls) {
+        int arity = kv.Key;
+
+        List<string> TyList(string prefix) {
+          var l = arity switch {
+            1 => new List<string> { prefix },
+            _ => Enumerable.Range(1, arity).Select(i => $"{prefix}{i}").ToList()
+          };
+          l.Add($"{prefix}Result");
+          return l;
+        }
+
+        string TPs(IEnumerable<string> l) => $"<{l.Comma()}>";
```

Comment:
It's about reusing, not explaining in this instance.

Reply:
I see. Is the purpose of such reuse then to improve maintainability, so that if say C#'s `<>` syntax is changed, we only need to update `TPs` and not all the locations where the `<>` string occurs? 

I think in the general case, nameless variables/methods that exist only for the purpose of reuse, do not improve maintainability. Suppose I have a widely used method called X and at some point I discover that one of the callers of X needs X to do something slightly different, so I change `X`. However, now I'm not sure whether the other callers should still call `X` or a copy of the old `X`. Without having name that explains what a thing does/is, it's difficult to know what the contents should be at any point in time.